Replaced the component callback to a connect call to free up comp init.

norunners opened the pull request:
Replaced the component callback to a connect call to free up component initialization.

Simplified components by embedding components which preserves component behavior.
Added read and write locking to the subscription map usage.
Chose Go idiomatic renaming and added interfaces.


I'm going to explain some changes I've made beyond the simplified commit message.

1. Replaced the component callback to a connect call to free up component initialization.
This was to remove the limitation of the component initialization only accepting a store as an parameter. I added a connect method which components pass access to the store. Therefore, components can be constructed however they want, with whatever passed down to them.

2. Simplified components by embedding components which preserves component behavior.
I simply changed composition to embedding to allow components to continue to satisfy the behavior they choose to, so that vstore usage does not hinder vecty usage.

3. Added read and write locking to the subscription map usage.
This is to protect vstore from race conditions against pubsub logic.

4. Chose Go idiomatic renaming and added interfaces.
This is just to try and be more Go spirited and break away from the JS world. I do not feel strongly about this change.

Additional comments.

1. I let all subscribed components be rerendered if the state changes. This was easy to do after using the StoreComponent interface and it relieves the user from thinking about rerendering while using vstore for state management.

2. I changed the examples state and dispatch callback fields to be the store, this is just preference, I generally stay away from function callbacks if it's trivial to use the behavior when needed, same with state. I do not feel strongly about this change.

3. I changed all the structs to unexpected and created interfaces to export, this is just to make both the users and developers more aware of the API vstore is providing for users, it can also ease unit testing if they satisfy the interfaces. I do not feel strongly about this change.

norunners committed:
Replaced the component callback to a connect call to free up component initialization.

Simplified components by embedding components which preserves component behavior.
Added read and write locking to the subscription map usage.
Chose Go idiomatic renaming and added interfaces.

norunners commented on store.go:
```diff
@@ -1,51 +1,70 @@
-// It also calls any subscription functions if exist.
-func (s *Store) Dispatch(action interface{}) {
+// NOTE: vecty.Rerender is called in a goroutine for each subscribed component.
+func (s *store) Dispatch(action interface{}) {
```
The dispatch logic needs to be in a goroutine to allow blocking calls, this does not effect the state from being updated and rerendered the same as before. I'll update the pull request.

marwan-at-work commented on store.go:
```diff
@@ -1,51 +1,70 @@
+	defer s.mtx.RUnlock()
+
+	for comp := range s.comps {
+		comp := comp
```
unnecessary reassign?

norunners replied:
It actually creates a new var here, since for range loops reuse the same pointer address per iteration. See parallelized table driven tests https://rakyll.org/parallelize-test-tables/ for an example.

marwan-at-work replied:
@norunners this is because `t.Run` does not take an extra parameter to register the changing loop variable. In our case, we are passing `comp` to `go vecty.Rerender()`. Which means, that `comp` is being registered because we are passing it as a function parameter and not as a closure. So this should be safe to remove. Check this wiki for more info: https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables

marwan-at-work commented on store.go:
```diff
@@ -1,51 +1,70 @@
+	r   Reducer
+	mws []Middleware
+
+	mtx   sync.RWMutex
```
why not a regular `sync.Mutex` since there's no reading or writing in the way we're syncing Dispatch.

norunners replied:
Subscribe and unsubscribe use the write lock and dispatch uses the read lock while rerendering. It may make sense to change dispatch to use a write lock on the reduce call.

marwan-at-work commented on comp.go:
```diff
@@ -2,47 +2,40 @@ package vstore
+// Mount subscribes the component to the store and optionally calls mount.
+func (c *storeComponent) Mount() {
+	c.store.sub(c)
+	if mounter, ok := c.StoreComponent.(vecty.Mounter); ok {
```
this is neat. However, we would have to support and keep up with all the life-cycle hooks that vecty will provide. Think react's `componentWillMount`, `componentShouldUpdate`, and `componentWillReceiveProps`. It would be an annoyance to have to keep an eye on every new vecty life-cycle feature. 

The simple workaround is this : 

```go
func (c *storeComponent) Render() vecty.HTML {
    return elem.Div(c.StoreComponent)
}
```

this will cause all the life cycle hooks of the child component to trigger. Note that we cannot do `return c.StoreComponent.Render()` so our higher order component has to wrap the child with an element.

pdf replied:
I haven't had a chance to provide examples there as I've been working on keyed children (now merged), but I filed gopherjs/vecty#148 exactly because of this sort of use-case.

marwan-at-work replied:
@pdf that would be ideal! As of now, the extra `div` doesn't bother me because it means `vstore` doesn't need to keep up with vecty's lifecycles and the user experience stays fluid.

pdf replied:
Introducing an additional (and unmodifiable!) DOM element can cause significant problems.  I think there's a general design problem here introduced by trying to follow the redux example too closely.

I don't see any reason to wrap components for this functionality, instead a new interface for components to implement that describes a callback for the store to hit would allow components to register themselves to the store, without needing to retain a reference to the store, and without needing to wrap components.

It's super-late here now, so I've got to get some sleep, will have to follow up in the coming days.

marwan-at-work replied:
@pdf you should define and give examples of what "significant problems" mean.

pdf replied:
Adding extra structural elements breaks layouts that rely on specific nesting or CSS.

marwan-at-work replied:
css is cascading, so it shouldn't be a problem as long as a user can specify what element the wrapping parent is, and can even pass it classes or props. which definitely should be included in vstore as a feature. 

All of that discussion, however, would be solved if that issue you mentioned initially is resolved on the vecty side. How likely is that @pdf ?

pdf replied:
It's not always so simple, particularly when dealing with third-party CSS frameworks and the like, and can be difficult to deal with when working with grids/flexbox/etc.

Changing the component render return would be a moderately complex change, so I'm not sure if it will happen or not at this stage.  I need to review the current issue priorities before I can investigate in more depth and work out if the cost (time/complexity) is worth it.

norunners replied:
Keep in mind we are using embedding not composition aka wrapping. This allows the component will continue to satisfy any behavior it previous did before the embedding. Therefore, we will not have to keep up with all the vecty lifecycle hooks, just the ones we decorate.

marwan-at-work replied:
@norunners as long as the rest of the lifecycle hooks get automatically triggered without having for vstore to keep up, then we're good 👍

marwan-at-work commented on comp.go:
```diff
@@ -2,47 +2,40 @@ package vstore
+// Unmount unsubscribes the component to the store and optionally calls unmount.
+func (c *storeComponent) Unmount() {
+	c.store.unsub(c)
+	if unmounter, ok := c.StoreComponent.(vecty.Unmounter); ok {
```
same as above.

norunners replied:
I'm replying to this comment because there is no response box below it.

"The last thing I want to point out, is that in redux a Store is generic and is not coupled at all with any UI framework like React, Backbone..etc. What makes react and redux coupled together is the react-redux module that gives us the Provider and connect functionalities."

My idea was to continue to allow use of vecty without being forced to use vstore, but also simplify vstore by being vecty aware. If you want to break this apart, we should provide two packages.

marwan-at-work commented on example/main.go:
```diff
@@ -12,41 +11,40 @@ import (
 	vecty.Core
-	Number   int `vecty:"prop"`
-	Dispatch func(action interface{})
+	store vstore.Store
```
One neat feature of react's component-structure is the ability to declare what data a component requires in order to be rendered correctly. This is done through `ReactProps` since JavaScript is not a statically typed language.

However, with Go, we can declare our data-dependencies in the struct Fields, which is even better since the properties are checked at compile time. Therefore, you can just look at a component `struct` and see exactly what kind of data it requires.

Therefore, `vstore` shouldn't take away this feature form the user. So we should put back the `Number` field here and populate it in the `Connect` function. This means that `Dispatch` will need to call `Connect` on every component before it calls `go vecty.Rerender(comp)`. 


This is even better than my original solution because we won't even need to add the `vecty:"prop"` struct tag anymore.

norunners replied:
I'd like the user to have the option to have the store or state as a component struct field. Do I need to change when connect is called for the case of the component having the state as a struct field?

marwan-at-work commented on example/main.go:
```diff
@@ -12,41 +11,40 @@ import (
 // NewMainComp is the mainComp constructor
-func NewMainComp(s *vstore.Store) vecty.Component {
-	state := s.GetState().(*store.State)
+func NewMainComp() vstore.StoreComponent {
```
this is not needed anymore. we can just pass `&mainComp{}` in the `body`'s `Render` method. We should also not pass the entire `store` anymore since we only use the `Dispatch` method, so we can just pass `Dispatch` to be explicit about what this component needs.

norunners replied:
Yes, it just demonstrates the user can have addition logic in the initialize method.

marwan-at-work commented on example/main.go:
```diff
@@ -12,41 +11,40 @@ import (
-		Number:   state.Number,
-		Dispatch: s.Dispatch,
-	}
+func (comp *mainComp) Connect(store vstore.Store) {
```
should be `(m *mainComp)` like the rest of the methods below. Or the other way around :)

norunners replied:
Yes, I'll keep it consistent.

marwan-at-work commented on example/main.go:
```diff
@@ -12,41 +11,40 @@ import (
-		Dispatch: s.Dispatch,
-	}
+func (comp *mainComp) Connect(store vstore.Store) {
+	comp.store = store
```
here we can just do 

```go
state := m.store.State().(*store.State)

comp.Dispatch = store.Dispatch
comp.Number = state.Number
```

norunners replied:
Yes, either way works, we should continue to provide both options.

marwan-at-work replied:
@norunners i tested this and it didn't work unless you call `connect` from the `dispatch` function. Can you double check? Thanks!

norunners replied:
I believe we should continue to call connect before rerender. Although it would work if we have the state reference as a component struct field instead of the number int but this assumes the state reference remains the same.

marwan-at-work reviewed (changes requested):
@norunners thanks a lot for the PR! This is an awesome improvement to the original design. 

I only have a couple of comments regarding type safety and component life cycles. 

The last thing I want to point out, is that in redux a `Store` is generic and is not coupled at all with any UI framework like React, Backbone..etc. What makes react and redux coupled together is the `react-redux` module that gives us the `Provider` and `connect` functionalities.

Therefore, I'm not sure if we should remove the `Subscribe` function from `*vstore.Store`. In redux, a dispatch just calls all the callbacks and passes them an instance of the state. Therefore, we can do the same thing and others can use `vstore` in other frameworks and not just `vecty`

On the other hand, the intention of this package is for vecty, so I'm a bit indifferent.

norunners committed:
Let dispatch run in a goroutine and connect components before rerendering.

norunners commented:
Here is a follow up commit addressing review comments.
1. Let dispatch run in a grouting.
2. Changed dispatch use a regular lock because state may be changed in a reduce call.
3. Call component connect before rerendering the component.
4. Reverted the example to have a dispatch function and a number state field.

https://github.com/marwan-at-work/vstore/pull/1/files/c6a525e0310a9f009cd4a01b8c5a2b972b1e7ae4..3e6fe79b480d9f6f6981068f5a9960e2b073bfc6

marwan-at-work commented on store.go:
```diff
@@ -1,51 +1,81 @@
-func (s *Store) GetState() interface{} {
-	return s.r
+// rerender connects the store to the component and calls vecty.Rerender on the component.
+func (s *store) rerender(comp *storeComponent) {
```
i don't think we need to make this function a `*store` receiver method. Typically, a pointer method exists to update its state, but in this case it's just doing two externally scoped functions. You can either have this be a regular function, or you can inline it on line 73 like so

```go
    for comp := range s.comps {
        go func(comp *storeComponent) {
            comp.Connect(s)
            vecty.Rerender(comp)
        }(comp)
    }
```

norunners replied:
It's a trivial unexpected method, inline go functions are okay until someone adds complexity.

marwan-at-work reviewed (changes requested):
All looks good. Thanks for the effort :) just one last nit change.

norunners commented:
I updated the latest commit, feel free to pull as a squashed merge commit.

marwan-at-work commented:
@norunners thank you! I'll try to visit the repo again and add some documentation and benchmarking.
